bundles/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/se/connection/ValloxIpConnector.java

```diff
@@ -0,0 +1,144 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.vallox.internal.se.connection;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.openhab.binding.vallox.internal.se.ValloxSEConstants;
+import org.openhab.binding.vallox.internal.se.configuration.ValloxSEConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link ValloxIpConnector} is creates TCP/IP connection to Vallox.
+ *
+ * @author Miika Jukka - Initial contribution
+ */
+@NonNullByDefault
+public class ValloxIpConnector extends ValloxBaseConnector {
+
+    private final Logger logger = LoggerFactory.getLogger(ValloxIpConnector.class);
+    private final Thread readerThread = new TelegramReader("Vallox Telegram Reader");
+    private Socket socket = new Socket();
+
+    public ValloxIpConnector() {
+        logger.debug("Ip connector initialized");
+    }
+
+    /**
+     * Connect to socket
+     */
+    @Override
+    public void connect(ValloxSEConfiguration config) throws IOException {
+        if (isConnected()) {
+            return;
+        }
+        buffer.clear();
+        panelNumber = config.getPanelAsByte();
+        socket = new Socket();
+        socket.connect(new InetSocketAddress(config.tcpHost, config.tcpPort), ValloxSEConstants.CONNECTION_TIMEOUT);
+        socket.setSoTimeout(0); // Don't fail if there's nothing to read. Machine is powered down.
+        inputStream = socket.getInputStream();
+        outputStream = socket.getOutputStream();
+        logger.debug("Connected to {}:{}", config.tcpHost, config.tcpPort);
+
+        startTelegramReaderThread();
+        startProcessorThreads();
+        connected.set(true);
+    }
+
+    /**
+     * Start threads for receiving and processing telegrams
+     */
+    private void startTelegramReaderThread() {
+        if (!readerThread.isAlive()) {
+            readerThread.setDaemon(true);
+            readerThread.start();
+        }
```

Comment:
If the thread died/finished, I don't think you can call `start` again...

Reply:
That is correct. I previously noticed `IllegalThreadStateException` and tried to handle it this way which is wrong. I realized that this is a design flaw in the change to pass connector from `ThingHandler`. When configuration changes the `thing` changes and in the ends threads are stopped and then starting them throws exception.

Reply:
With quite simple changes I switched to dedicated `ThreadPool` and `Future` tasks. Let me test and i'll push a commit.